spikeinterface/toolkit/postprocessing/spike_locations.py

```diff
@@ -0,0 +1,180 @@
+                             get_template_extremum_channel_peak_shift)
+
+
+class SpikeLocationsCalculator(BaseWaveformExtractorExtension):
+    """
+    Localize spikes in 2D or 3D depending the method.
+
+    When a probe is 2D then:
+       * X is axis 0 of the probe
+       * Y is axis 1 of the probe
+       * Z is orthogonal to the plane of the probe
+
+    Parameters
+    ----------
+    waveform_extractor: WaveformExtractor
+        The waveform extractor object
+    ms_before: float
+        The left window, before a peak, in milliseconds.
+    ms_after: float
+        The right window, after a peak, in milliseconds.
+    method: 'center_of_mass' or 'monopolar_triangulation'
+        Method to use.
+    method_kwargs: dict of kwargs method
+        Keyword arguments for the chosen method:
+            'center_of_mass':
+                * local_radius_um: float
+                    For channel sparsity.
+            'monopolar_triangulation':
+                * local_radius_um: float
+                    For channel sparsity.
+                * max_distance_um: float, default: 1000
+                    Boundary for distance estimation.
+    {}
+
+    Returns
+    -------
+    spike_locations: np.array
+        The spike locations.
+            - If 'concatenated' all locations for all spikes and all units are concatenated
+    """    
+    extension_name = 'spike_locations'
+    
+    def __init__(self, waveform_extractor):
+        BaseWaveformExtractorExtension.__init__(self, waveform_extractor)
+
+        self._locations = None
+        self._all_spikes = None
+
+    def _set_params(self, ms_before=1, ms_after=1, method='center_of_mass',
+                    method_kwargs={}):
+
+        params = dict(ms_before=ms_before,
+                      ms_after=ms_after,
+                      method=method,
+                      method_kwargs=method_kwargs)
+        return params        
+        
+    def _specific_load_from_folder(self):
+        recording = self.waveform_extractor.recording
+        sorting = self.waveform_extractor.sorting
+
+        all_spikes = sorting.get_all_spike_trains(outputs='unit_index')
+        self._all_spikes = all_spikes
+
+        self._locations = []
+        for segment_index in range(recording.get_num_segments()):
+            file_locs = self.extension_folder / f'locations_segment_{segment_index}.npy'
+            locs_seg = np.load(file_locs)
+            self._locations.append(locs_seg)
+
+    def _reset(self):
+        self._locations = None
+    
+    def _specific_select_units(self, unit_ids, new_waveforms_folder):
+        # load filter and save amplitude files
+        for seg_index in range(self.waveform_extractor.recording.get_num_segments()):
+            loc_file_name = f"locations_segment_{seg_index}.npy"
+            locs = np.load(self.extension_folder / loc_file_name)
+            _, all_labels = self.waveform_extractor.sorting.get_all_spike_trains()[seg_index]
```

Comment:
This `self.waveform_extractor.sorting.get_all_spike_trains()` have a cost.
I would use the self._all_spikes and play with `keep_unit_indices` instead of `unit_ids` for the in1d.